xcvm/lib/core/src/asset.rs

```diff
@@ -0,0 +1,262 @@
+use crate::abstraction::IndexOf;
+use alloc::{collections::BTreeMap, string::ToString};
+use codec::{Decode, Encode};
+use core::ops::Add;
+use fixed::{types::extra::U16, FixedU128};
+use num::Zero;
+use scale_info::TypeInfo;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+/// Newtype for XCVM assets ID. Must be unique for each asset and must never change.
+/// This ID is an opaque, arbitrary type from the XCVM protocol and no assumption must be made on
+/// how it is computed.
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[derive(
+	Copy,
+	Clone,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Debug,
+	Encode,
+	Decode,
+	TypeInfo,
+	Serialize,
+	Deserialize,
+)]
+#[repr(transparent)]
+pub struct AssetId(pub u128);
+
+impl From<AssetId> for u128 {
+	fn from(val: AssetId) -> Self {
+		val.0
+	}
+}
+
+impl From<u128> for AssetId {
+	fn from(asset: u128) -> Self {
+		AssetId(asset)
+	}
+}
+
+impl From<PICA> for AssetId {
+	fn from(_: PICA) -> Self {
+		PICA::ID
+	}
+}
+
+impl From<ETH> for AssetId {
+	fn from(_: ETH) -> Self {
+		ETH::ID
+	}
+}
+
+impl From<USDT> for AssetId {
+	fn from(_: USDT) -> Self {
+		USDT::ID
+	}
+}
+
+impl From<USDC> for AssetId {
+	fn from(_: USDC) -> Self {
+		USDC::ID
+	}
+}
+
+pub struct InvalidAsset;
+pub struct PICA;
+pub struct ETH;
+pub struct USDT;
+pub struct USDC;
+
+/// List of XCVM compatible assets.
+/// /!\ The order matters and must not be changed, adding a network on the right is safe.
+pub type Assets = (InvalidAsset, (PICA, (ETH, (USDT, (USDC, ())))));
```

Comment:
Macro to make this

```
assets![InvalidAsset, PICA, ETH, USDT, USDC]?
```

Reply:
Not worth the overhead of instroducing an extra macro tbh, that will invalidate most IDE typechecking hints also